### PR TITLE
[PATCH v7] test: performance: ipsec: introduce burst processing

### PR DESCRIPTION
**Pull request by anoobj**
Introduce an option to prepare and submit multiple packets for IPsec
processing.

ODP allows multi APIs with packet alloc, free & IPsec processing. Usage
of these APIs would help in improving performance with burst cases. New
option '--burst' or '-b' can be used to enable burst processing. When
specified without any arguments, burst size would be set as 32. User can
control burst size by specifying the optional argument.

Usage: 'odp_ipsec --burst=512'

Signed-off-by: Anoob Joseph <anoobj@marvell.com>

### CODE REVIEW

File: test/performance/odp_ipsec.c [hunk 1]
```diff
@@ -139,6 +139,13 @@ typedef struct {
 	 * Specified through -u argument.
 	 */
 	int ah;
+
+	/*
+	 * Burst size.
+	 * Prepare and submit as many packets for IPsec processing in each
+	 * iteration of the loop.
+	 */
+	int burst_sz;
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:148]: Maybe burst_size would be better name than burst_sz in all places where burst_sz is used?

File: test/performance/odp_ipsec.c [hunk 2]
```diff
@@ -584,28 +591,29 @@ static uint8_t test_data[] = {
...
-		return ODP_PACKET_INVALID;
-
-	pkt = odp_packet_alloc(pkt_pool, payload_length);
-	if (pkt == ODP_PACKET_INVALID) {
-		ODPH_ERR("failed to allocate buffer\n");
-		return pkt;
+	ret = odp_packet_alloc_multi(pkt_pool, payload_length, pkt, num);
+	if (ret != num) {
+		ODPH_ERR("Could not allocate buffer\n");
+		odp_packet_free_sp(pkt, ret);
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:603]: ret may be negative here. This should be called only if it is positive.

File: test/performance/odp_ipsec.c [hunk 3]
```diff
@@ -643,41 +655,58 @@ run_measure_one(ipsec_args_t *cargs,
 
 	while ((packets_sent < cargs->iteration_count) ||
 	       (packets_received < cargs->iteration_count)) {
-		if ((packets_sent < cargs->iteration_count) &&
+		int num_out = cargs->iteration_count - packets_sent;
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:658]: Looking at the (packets_sent < cargs->iteration_count) condition, now the iteration_count is really packet count and not the number of iterations or number of times the IPsec enqueue API gets called. Maybe the description of -i in usage() should be updated?

File: test/performance/odp_ipsec.c [hunk 4]
```diff
@@ -643,41 +655,58 @@ run_measure_one(ipsec_args_t *cargs,
...
 				break;
 			}
-			if (odp_packet_has_error(out_pkt)) {
-				odp_ipsec_packet_result_t result;
 
-				odp_ipsec_result(&result, out_pkt);
-				ODPH_ERR("Received error packet: %d\n",
-					 result.status.error.all);
+			for (i = 0; i < num_out; i++) {
+				if (odp_packet_has_error(out_pkt[i])) {
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:692]: I think checking odp_packet_has_error() here and not checking the IPsec packet result here and in the async function is wrong. But maybe that needs to be fixed in a different patch.

File: test/performance/odp_ipsec.c [hunk 5]
```diff
@@ -726,52 +760,63 @@ run_measure_one_async(ipsec_args_t *cargs,
...
+		 i = 0;
+
 		if (cargs->schedule)
 			ev = odp_schedule(NULL,
 					  ODP_SCHED_NO_WAIT);
 		else
 			ev = odp_queue_deq(out_queue);
 
 		while (ev != ODP_EVENT_INVALID) {
-			odp_packet_t out_pkt;
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:758]: Maybe this async test loop should use odp_schedule_multi() now?

File: test/performance/odp_ipsec.c (hunk 4 above)
Comment (JannePeltonen) [reply test/performance/odp_ipsec.c:692]: I made PR 1506 to fix this.

File: test/performance/odp_ipsec.c (hunk 5 above)
Comment (JannePeltonen) [reply test/performance/odp_ipsec.c:758]: PR 1506 takes odp_schedule_multi and odp_queue_deq_multi into use here. Please take a look.

File: test/performance/odp_ipsec.c (hunk 1 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:148]: Agreed. Will make this change in v2.

File: test/performance/odp_ipsec.c (hunk 2 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:603]: Agreed. Will make this change in v2.

File: test/performance/odp_ipsec.c (hunk 3 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:658]: Agreed. Do you suggest renaming the variable as well? (like, iteration_count -> packet_count?)

Or would updating description of -i option good enough?

Comment (JannePeltonen) [reply test/performance/odp_ipsec.c:658]: Its probably good to rename the variable too for clarity.

File: test/performance/odp_ipsec.c [hunk 6]
```diff
@@ -643,41 +655,58 @@ run_measure_one(ipsec_args_t *cargs,
...
 	       (packets_received < cargs->iteration_count)) {
-		if ((packets_sent < cargs->iteration_count) &&
+		int num_out = cargs->iteration_count - packets_sent;
+
+		num_out = num_out > burst_sz ? burst_sz : num_out;
+
+		if (num_out &&
 		    (packets_sent - packets_received <
 		     cargs->in_flight)) {
-			odp_packet_t out_pkt;
```
[the remark this answers is not in the document]
Comment (anoobj) [reply test/performance/odp_ipsec.c:649]: I think it's better to have this fixed in code. I'll address this in v2.

**Comment by anoobj** [comment]
@JannePeltonen Do you have any comments on v2? If not I can push next version with your patch from PR 1506

File: test/performance/odp_ipsec.c [hunk 7]
```diff
@@ -971,8 +1031,18 @@ static void parse_args(int argc, char *argv[], ipsec_args_t *cargs)
 		case 'd':
 			cargs->debug_packets = 1;
 			break;
-		case 'i':
-			cargs->iteration_count = atoi(optarg);
+		case 'c':
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:1034]: Now that the argument letter changes, odp_ipsec_run.sh needs an update too.

File: test/performance/odp_ipsec.c [hunk 8]
```diff
@@ -641,43 +655,64 @@ run_measure_one(ipsec_args_t *cargs,
...
+
+		/* Enqueue up to (max in flight - current in flight) */
+		in_flight = packets_sent - packets_received;
+		pkts_allowed = max_in_flight - in_flight;
+		num_out = num_out > pkts_allowed ? pkts_allowed : num_out;
 
-			pkt = make_packet(pkt_pool, payload_length);
-			if (ODP_PACKET_INVALID == pkt)
+		if (odp_likely(num_out)) {
+			odp_packet_t out_pkt[POOL_NUM_PKT];
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:671]: This could be out_pkt[num_out].

File: test/performance/odp_ipsec.c [hunk 9]
```diff
@@ -693,11 +728,14 @@ run_measure_one_async(ipsec_args_t *cargs,
...
+	int in_flight, pkts_allowed, i, num_pkts, rc = 0;
+	const int max_in_flight = cargs->in_flight;
+	const int burst_size = cargs->burst_size;
 	odp_ipsec_out_param_t param;
 	odp_pool_t pkt_pool;
 	odp_queue_t out_queue;
-	odp_packet_t pkt = ODP_PACKET_INVALID;
-	int rc = 0;
+	odp_packet_t pkt_out[POOL_NUM_PKT];
+	odp_packet_t pkt[POOL_NUM_PKT];
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:738]: These could also be smaller. If moved down in the loop, then num_pkts as the size would suffice, it seems.

Comment (JannePeltonen) [reply test/performance/odp_ipsec.c:738]: ...or max_in_flight.

File: test/performance/odp_ipsec.c [hunk 10]
```diff
@@ -641,43 +655,64 @@ run_measure_one(ipsec_args_t *cargs,
...
+				for (i = 0; i < num_out; i++)
+					odp_packet_print_data(pkt[i], 0,
+							      odp_packet_len(pkt[i]));
+			}
 
-			rc = odp_ipsec_out(&pkt, 1,
-					   &out_pkt, &num_out,
+			i = num_out;
+
+			rc = odp_ipsec_out(pkt, num_out,
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:688]: If this does not consume all packets, then the leftover packet are leaked.

File: test/performance/odp_ipsec.c [hunk 11]
```diff
@@ -722,56 +763,72 @@ run_measure_one_async(ipsec_args_t *cargs,
...
-				odp_packet_print_data(pkt, 0,
-						      odp_packet_len(pkt));
+			if (cargs->debug_packets) {
+				for (i = 0; i < num_pkts; i++)
+					odp_packet_print_data(pkt[i], 0,
+							      odp_packet_len(pkt[i]));
+			}
 
-			rc = odp_ipsec_out_enq(&pkt, 1,
+			rc = odp_ipsec_out_enq(pkt, num_pkts,
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:793]: There is a packet leak also here if the enqueue does not consume all input packets.

File: test/performance/odp_ipsec.c (hunk 8 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:671]: Agreed. Will update in v3.

File: test/performance/odp_ipsec.c (hunk 10 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:688]: Agreed. Will fix in v3.

File: test/performance/odp_ipsec.c (hunk 11 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:793]: Agreed. Will fix in v3.

File: test/performance/odp_ipsec.c (hunk 9 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:738]: Agreed. Will move it deeper and use max_in_flight & num_pkts.

File: test/performance/odp_ipsec.c (hunk 7 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:1043]: Updated in v3.

**Comment by anoobj** [comment]
@JannePeltonen Do you have any further comments?

File: test/performance/odp_ipsec.c [hunk 12]
```diff
@@ -632,17 +642,22 @@ run_measure_one(ipsec_args_t *cargs,
 		time_record_t *start,
 		time_record_t *end)
 {
+	int in_flight, pkts_allowed, num_out, num_enq, rc = 0;
+	const int max_in_flight = cargs->in_flight;
+	const int burst_size = cargs->burst_size;
+	odp_packet_t pkt[POOL_NUM_PKT];
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:648]: This could be made much smaller if moved down, i.e. similar change could be done here as was now done in the async case.

File: test/performance/odp_ipsec.c [hunk 13]
```diff
@@ -654,38 +669,57 @@ run_measure_one(ipsec_args_t *cargs,
...
+				for (i = 0; i < num_enq; i++)
+					odp_packet_print_data(pkt[i], 0,
+							      odp_packet_len(pkt[i]));
+			}
+
+			num_out = num_enq;
 
-			rc = odp_ipsec_out(&pkt, 1,
-					   &out_pkt, &num_out,
+			rc = odp_ipsec_out(pkt, num_enq,
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:702]: Leftover packets are leaked if this does not consume all packets.

File: test/performance/odp_ipsec.c [hunk 14]
```diff
@@ -654,38 +669,57 @@ run_measure_one(ipsec_args_t *cargs,
...
-				odp_packet_print_data(out_pkt, 0,
-						      odp_packet_len(out_pkt));
-			odp_packet_free(out_pkt);
+			if (cargs->debug_packets) {
+				for (i = 0; i < num_out; i++)
+					odp_packet_print_data(out_pkt[i], 0,
+							      odp_packet_len(out_pkt[i]));
+			}
+
+			odp_packet_free_sp(out_pkt, num_enq);
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:722]: This can try to free invalid or stale packet handles. The number of packets in the array is not num_enq but num_out, which may be less than num_enq.

File: test/performance/odp_ipsec.c [hunk 15]
```diff
@@ -730,55 +768,76 @@ run_measure_one_async(ipsec_args_t *cargs,
...
-			if (ODP_PACKET_INVALID == pkt)
+		num_pkts = cargs->packet_count - packets_sent;
+
+		/* Enqueue up to burst size */
+		num_pkts = num_pkts > burst_size ? burst_size : num_pkts;
+
+		/* Enqueue up to (max in flight - current in flight) */
+		in_flight = packets_sent - packets_received;
+		pkts_allowed = max_in_flight - in_flight;
+		num_pkts = num_pkts > pkts_allowed ? pkts_allowed : num_pkts;
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:784]: I think this should be:

num_pkts = num_pkts > pkts_allowed ? 0 : num_pkts;

This would make the burst size always (except for the last burst) the same what was specified in the -b option, which is what one would expect. Now the burst size can get smaller if IPsec processing is slower than the test loop (as is expected) and then using a smaller burst size can further slow down the processing so that the loop ends up using small burst size for pretty much the whole test.

And the same comment for the sync case.

With this change (always using the requested burst size), I think the program should complain and exit if the in_flight parameter is less than the burst size parameter.

File: test/performance/odp_ipsec.c [hunk 16]
```diff
@@ -632,17 +642,22 @@ run_measure_one(ipsec_args_t *cargs,
 		time_record_t *start,
 		time_record_t *end)
 {
+	int in_flight, pkts_allowed, num_out, num_enq, rc = 0;
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:645]: I think num_enq is not a very good name here in the sync case which does not really enqueue packets. The async case seems to use num_pkts which could be used here too (and it would have the added benefit of making the two functions more similar in variable naming).

File: test/performance/odp_ipsec_run.sh [hunk 17]
```diff
@@ -7,9 +7,9 @@
 
 TEST_DIR="${TEST_DIR:-$(dirname $0)}"
 
-# Run with a small number of iterations in make check
+# Run with small number of packets in make check
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec_run.sh:10]: There is no reason to remove the indefinite article.

File: test/performance/odp_ipsec.c (hunk 14 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:722]: If out_pkt is replaced with pkt, that should be enough, right?

File: test/performance/odp_ipsec.c (hunk 12 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:648]: Agreed. Will change in v4.

File: test/performance/odp_ipsec.c (hunk 14 above)
Comment (JannePeltonen) [reply test/performance/odp_ipsec.c:722]: Yes.

File: test/performance/odp_ipsec.c (hunk 16 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:645]: No strong opinion on variable name. Will rename as part of v4.

File: test/performance/odp_ipsec.c (hunk 15 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:784]: In that case, are you okay with making packet_count a multiple of burst_size? Then we can improve/simplify the run_measure_*() loops.

File: test/performance/odp_ipsec.c (hunk 14 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:722]: I'll make this change as part of v4.

File: test/performance/odp_ipsec.c (hunk 15 above)
Comment (JannePeltonen) [reply test/performance/odp_ipsec.c:784]: Maybe making packet_count a multiple of burst size does not quite enable the simplification that you may be thinking, because the ipsec out/enq function can still consume only part of the burst, making the count of processed packets not divisible by the burst size.

But simplification is good, and I think the test loop could just stop when the remaining packet count falls below the burst count (or below zero or whatever comes naturally). IOW, the packet count would not be an exact number anymore (because who would actually care?), but just a limit to stop the test at that point.

Actually, I think it would be even better if the test termination would not be based on packet count but elapsed time, as that would make it easier to run the test loop in multiple threads simultaneously without syncing about the number of processed packets. That is what I have been doing with my private test apps (i.e. time based termination & multiple threads running a test loop).

But since this test app does not support multiple threads, packet count limit is ok. I just do not see any need for it to be exact. So a loop that always sends a full burst or nothing and then terminates roughly, but not precisely, after the requested number of packets should be fine. And it would simpler and clearer than having extra logic for handling a partial burst at the end.

Comment (anoobj) [reply test/performance/odp_ipsec.c:784]: @JannePeltonen Thanks for your inputs. I hadn't thought about partial enqueue of the burst. So, agreed that such a change won't lead to the simplification I had in mind. So, I'll stick to your earlier suggestion.

File: test/performance/odp_ipsec.c (hunk 14 above)
Comment (JannePeltonen) [reply test/performance/odp_ipsec.c]: Sorry, I misread the question when I answered. Replacing num_enq with out_pkt would have fixed the problem. Replacing out_pkt with pkt is wrong.

odp_ipsec_out() consumes some or all of the input packets and produces output packets. The output packets are not necessarily the same as the input packets although they can be. So after odp_ipsec_out() call you have a zero or more packets left in the input array and you need to free them in order not to leak them. You also have zero or more packets in the output array, but they are new packets and cannot be freed using the old input array handles that are now stale (except that in some implementations, like linux-gen, they may sometimes (but not always), be the same handles in which case the current code accidentally works).

File: test/performance/odp_ipsec.c (hunk 15 above)
Comment (JannePeltonen) [reply test/performance/odp_ipsec.c]: > With this change (always using the requested burst size), I think the program should complain and exit if the in_flight parameter is less than the burst size parameter.

I think something like this should be done or otherwise the program gets stuck when run with smaller -f than -b.

Comment (anoobj) [reply test/performance/odp_ipsec.c]: Agreed. I'll have this addressed in v5.

File: test/performance/odp_ipsec.c (hunk 14 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c]: Agreed. Will have this changed in v5.

File: test/performance/odp_ipsec_run.sh (hunk 17 above)
Comment (JannePeltonen) [reply test/performance/odp_ipsec_run.sh:10]: It disappeared again in v5.

File: test/performance/odp_ipsec.c [hunk 18]
```diff
@@ -654,38 +668,62 @@ run_measure_one(ipsec_args_t *cargs,
...
-						      odp_packet_len(out_pkt));
-			odp_packet_free(out_pkt);
+			if (cargs->debug_packets) {
+				for (i = 0; i < num_out; i++)
+					odp_packet_print_data(out_pkt[i], 0,
+							      odp_packet_len(out_pkt[i]));
+			}
+
+			if (odp_unlikely(num_out != num_pkts))
+				odp_packet_free_sp(&pkt[num_out], num_pkts - num_out);
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:725]: This code makes the assumption that the number of consumed packets (rc) is the same as the number of output packets (num_out), which won't be true as soon as fragmentation offload gets enabled. It would be more robust to check if the number of consumed packets (i.e. rc) is not the same as the number of input packets and also use the number of consumed packet instead of the number of output packets in the freeing statement. The number of output packets is not directly relevant to freeing the unconsumed packets.

File: test/performance/odp_ipsec.c [hunk 19]
```diff
@@ -1008,6 +1086,11 @@ static void parse_args(int argc, char *argv[], ipsec_args_t *cargs)
 		}
 	}
 
+	if (cargs->in_flight < cargs->burst_size) {
+		printf("-f (flight) should be greater than -b (burst)\n");
```
Comment (JannePeltonen) [inline test/performance/odp_ipsec.c:1090]: Burst size should not exceed the maximum number of in-flight packets, but it may equal it.

File: test/performance/odp_ipsec.c (hunk 18 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:725]: @JannePeltonen So I'll replace num_out with rc in this code snippet and that should be good enough, right?

File: test/performance/odp_ipsec.c (hunk 19 above)
Comment (anoobj) [reply test/performance/odp_ipsec.c:1090]: @JannePeltonen  Shall I update the text as,

`printf("-f (flight) must be greater than or equal to -b (burst)\n");`

Or did you mean use the above text verbatim?

File: test/performance/odp_ipsec.c (hunk 18 above)
Comment (JannePeltonen) [reply test/performance/odp_ipsec.c:725]: Yes, that should fix it. There is already that kind of code in the async case, so this just becomes more similar.

File: test/performance/odp_ipsec.c (hunk 19 above)
Comment (JannePeltonen) [reply test/performance/odp_ipsec.c:1090]: Actually, your proposed update looks ok whereas copying my comment verbatim would make the error message more complicated than it needs to be, I think.

**Review by JannePeltonen — APPROVED** [review]
Looks good.
